tests/optional.cpp

```diff
@@ -316,6 +316,64 @@ TEST_CASE("optional::if_set()", "[optional]")
         REQUIRE(i == 3);
         REQUIRE(i2 == 4);
     }
+
+    SECTION("argument forwarding")
+    {
+        optional o { std::make_unique<std::string>("1") };
+        auto test_argument = 2;
+
+        auto& result = o.if_set([](auto& value_ptr, auto& test_argument) 
+        { 
+            REQUIRE(*value_ptr == "1");
+            REQUIRE(test_argument == 2); 
+        }, test_argument);
+
+        // if_set returns an optional - either the return value of the functor
+        // or, if the functor returns void, it will return itself
+        REQUIRE(&result == &o);
+
+        // move semantics allow us to move the value out
+        auto new_o = std::move(o).if_set([](auto value_ptr, auto& test_argument) 
+        { 
+            REQUIRE(test_argument == 2);
+            return std::make_unique<std::string>(*value_ptr + "_2"); 
+        }, test_argument);
+
+        // o must have been set so new_o also
+        REQUIRE_FALSE(new_o.empty());
+        new_o.if_set([](auto& value_ptr, auto& test_argument) 
+        { 
+            REQUIRE(*value_ptr == "1_2"); 
+            REQUIRE(test_argument == 2);
+        }, test_argument);
+
+        // check the const method of if_set works, also check that we can pass move only parameters
+        const optional const_o { 4 };
+        const auto& new_const_o = const_o.if_set([](auto& value, auto test_pointer)
+        {
+            REQUIRE(value == 4);
+            REQUIRE(*test_pointer == 10);
+            return value;
+        }, std::move(std::make_unique<int>(10)));
```

Comment:
this move here is unnecessary, as std::make_unique's return value is implicitly an r-value